File: solutions/problem_1471_the_k_strongest_values_in_an_array/solution.cpp
/*
 * Problem 1471: The k Strongest Values in an Array
 * ==============================================
 * Difficulty: Medium
 * Tags: Array, Two Pointers, Sorting
 * Pattern: Two Pointers on Sorted Array
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(1)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> getStrongest(vector<int>& arr, int k) {
        // Sort + two pointers - O(n log n) time
        sort(arr.begin(), arr.end());
        int left = 0, right = arr.size() - 1;
        while (left < right) {
            int curr = arr[left] + arr[right];
            if (curr < k) {
                left++;
            } else {
                right--;
            }
        }
        return {};
    }
};

File: solutions/problem_1471_the_k_strongest_values_in_an_array/solution.py
"""
Problem 1471: The k Strongest Values in an Array
==============================================
Difficulty: Medium
Tags: Array, Two Pointers, Sorting
Pattern: Two Pointers on Sorted Array

Time Complexity:  O(n log n)
Space Complexity: O(1)
"""

class Solution:
    def getStrongest(self, arr: List[int], k: int) -> List[int]:
        # Sort + two pointers - O(n log n) time
        arr.sort()
        left, right = 0, len(arr) - 1
        result = []
        while left < right:
            curr_sum = arr[left] + arr[right]
            if curr_sum < k if isinstance(k, int) else 0:
                left += 1
            else:
                right -= 1
        return result
